File: src/org/epistasis/snpgen/ui/ModelCreationTable.java
package org.epistasis.snpgen.ui;

import java.awt.Color;
import java.awt.Component;
import java.awt.event.FocusEvent;
import java.awt.event.FocusListener;
import java.util.ArrayList;
import java.util.Arrays;

import javax.swing.AbstractCellEditor;
import javax.swing.BoxLayout;
import javax.swing.JPanel;
import javax.swing.JTable;
import javax.swing.JTextField;
import javax.swing.event.TableModelEvent;
import javax.swing.table.DefaultTableModel;
import javax.swing.table.TableCellEditor;
import javax.swing.table.TableCellRenderer;
import javax.swing.table.TableModel;

public class ModelCreationTable extends JTable {
	private static final long serialVersionUID = 1L;

	private static final int kColumnWidgetCount = 2;

	DefaultTableModel tableModel;
	ColumnWidget[] widgets = null;

	public ModelCreationTable(final DefaultTableModel inTableModel) {
		super(inTableModel);

		tableModel = inTableModel;
		widgets = new ColumnWidget[ModelCreationTable.kColumnWidgetCount];
		Arrays.fill(widgets, null);
		for (int i = 0; i < ModelCreationTable.kColumnWidgetCount; ++i) {
			setWidgetForColumn(tableModel, i);
		}

		adjustRows();
		setShowGrid(true);
		setGridColor(Color.black);
		setTableHeader(createDefaultTableHeader());
	}

	// public static ModelCreationTable createModelCreationTable()
	// {
	// DefaultTableModel model = new DefaultTableModel();
	//
	// model.addColumn("SNP");
	// model.addColumn("Minor Allele Frequency");
	//
	// ModelCreationTable table = new ModelCreationTable(model);
	// return table;
	// }

	public TableModel getTableModel() {
		return getModel();
	}

	@Override
	public void tableChanged(final TableModelEvent inEvent) {
		super.tableChanged(inEvent);
		adjustRows();
	}

	private void adjustRows() {
		int height;

		if (widgets != null) {
			for (int i = 0; i < ModelCreationTable.kColumnWidgetCount; ++i) {
				if (widgets[i] != null) {
					widgets[i].adjustRowCount(getRowCount());
				}
			}
			for (int row = 0; row < getRowCount(); ++row) {
				height = getPreferredRowHeight(row);
				setRowHeight(row, height);
			}
		}
	}

	private int getPreferredRowHeight(final int inRow) {
		int height, maxHeight;

		maxHeight = getRowHeight();
		if (widgets != null) {
			for (int i = 0; i < ModelCreationTable.kColumnWidgetCount; ++i) {
				if (widgets[i] != null) {
					height = widgets[i].getPreferredRowHeight(inRow);
					if (height > maxHeight) {
						maxHeight = height;
					}
				}
			}
		}
		return maxHeight;
	}

	private ColumnWidget setWidgetForColumn(final TableModel inTableModel, final int inColumn) {
		final ColumnWidget widget = new ColumnWidget(inTableModel, inColumn);
		getColumnModel().getColumn(inColumn).setCellRenderer(widget);
		getColumnModel().getColumn(inColumn).setCellEditor(widget);
		widgets[inColumn] = widget;
		return widget;
	}

	public class ColumnWidget extends AbstractCellEditor implements TableCellRenderer, TableCellEditor {
		private static final long serialVersionUID = 1L;
		int column;
		TableModel tableModel;
		ArrayList<Panel> panelArray;

		public ColumnWidget(final TableModel inModel, final int inColumn) {
			column = inColumn;
			tableModel = inModel;
			// ArrayList<SnpGenDocument.DocModel> modelList =
			// tableModel.getModelList();
			// int modelCount = modelList.size();
			final int modelCount = inModel.getRowCount();
			panelArray = new ArrayList<Panel>(modelCount);
			for (int row = 0; row < modelCount; ++row) {
				panelArray.add(new Panel(row, inColumn, tableModel, tableModel.getValueAt(row, inColumn)));
			}
		}

		public void adjustRowCount(final int inRowCount) {
			for (int row = panelArray.size() - 1; row >= inRowCount; --row) {
				panelArray.remove(row);
			}
			for (int row = panelArray.size(); row < inRowCount; ++row) {
				panelArray.add(new Panel(row, column, tableModel, tableModel.getValueAt(row, column)));
			}
		}

		// Implement the one CellEditor method that AbstractCellEditor doesn't.
		@Override
		public Object getCellEditorValue() {
			// return currentColor;
			return new Integer(1);
		}

		public int getPreferredRowHeight(final int inRow) {
			return panelArray.get(inRow).getPreferredHeight();
		}

		// Implement the one method defined by TableCellEditor.
		@Override
		public Component getTableCellEditorComponent(final JTable table, final Object value, final boolean isSelected, final int row,
				final int column) {
			return panelArray.get(row);
		}

		@Override
		public Component getTableCellRendererComponent(final JTable table, final Object color, final boolean isSelected,
				final boolean hasFocus, final int row, final int column) {
			return panelArray.get(row);
		}

		public class Panel extends JPanel implements FocusListener {
			private static final long serialVersionUID = 1L;
			private JTextField textField;
			private final int row;
			private final int column;
			private final TableModel tableModel;

			public Panel(final int inRow, final int inColumn, final TableModel inTableModel) {
				row = inRow;
				column = inColumn;
				tableModel = inTableModel;
				setLayout(new BoxLayout(this, BoxLayout.Y_AXIS));
				setOpaque(true); // Must do this for background to show up.
				setBackground(Color.white);
				add(textField = new JTextField(3));
				textField.addFocusListener(this);
			}

			public Panel(final int inRow, final int inColumn, final TableModel inTableModel, final Object inValue) {
				this(inRow, inColumn, inTableModel);
				textField.setText(inValue.toString());
			}

			@Override
			public void focusGained(final FocusEvent e) {
			}

			@Override
			public void focusLost(final FocusEvent e) {
				tableModel.setValueAt(textField.getText(), row, column);
			}

			public int getPreferredHeight() {
				return getPreferredSize().height;
			}
		}
	}
}

File: src/gametes/display/ModelCreationTable.py
package org.epistasis.snpgen.ui;

import java.awt.Color;
import java.awt.Component;
import java.awt.event.FocusEvent;
import java.awt.event.FocusListener;
import java.util.ArrayList;
import java.util.Arrays;
import javax.swing.AbstractCellEditor;
import javax.swing.BoxLayout;
import javax.swing.JPanel;
import javax.swing.JTable;
import javax.swing.JTextField;
import javax.swing.event.TableModelEvent;
import javax.swing.table.DefaultTableModel;
import javax.swing.table.TableCellEditor;
import javax.swing.table.TableCellRenderer;
import javax.swing.table.TableModel;

public class ModelCreationTable extends JTable {
   private static final int kColumnWidgetCount = 2;
   DefaultTableModel tableModel;
   ModelCreationTable.ColumnWidget[] widgets = null;

   public ModelCreationTable(DefaultTableModel inTableModel) {
      super(inTableModel);
      this.tableModel = inTableModel;
      this.widgets = new ModelCreationTable.ColumnWidget[2];
      Arrays.fill(this.widgets, (Object)null);

      for(int i = 0; i < 2; ++i) {
         this.setWidgetForColumn(this.tableModel, i);
      }

      this.adjustRows();
      this.setShowGrid(true);
      this.setGridColor(Color.black);
      this.setTableHeader(this.createDefaultTableHeader());
   }

   public TableModel getTableModel() {
      return this.getModel();
   }

   private ModelCreationTable.ColumnWidget setWidgetForColumn(TableModel inTableModel, int inColumn) {
      ModelCreationTable.ColumnWidget widget = new ModelCreationTable.ColumnWidget(inTableModel, inColumn);
      this.getColumnModel().getColumn(inColumn).setCellRenderer(widget);
      this.getColumnModel().getColumn(inColumn).setCellEditor(widget);
      this.widgets[inColumn] = widget;
      return widget;
   }

   public void tableChanged(TableModelEvent inEvent) {
      super.tableChanged(inEvent);
      this.adjustRows();
   }

   private void adjustRows() {
      if (this.widgets != null) {
         int row;
         for(row = 0; row < 2; ++row) {
            if (this.widgets[row] != null) {
               this.widgets[row].adjustRowCount(this.getRowCount());
            }
         }

         for(row = 0; row < this.getRowCount(); ++row) {
            int height = this.getPreferredRowHeight(row);
            this.setRowHeight(row, height);
         }
      }

   }

   private void adjustRowHeight(int inRow) {
      int height = this.getPreferredRowHeight(inRow);
      this.setRowHeight(inRow, height);
   }

   private int getPreferredRowHeight(int inRow) {
      int maxHeight = this.getRowHeight();
      if (this.widgets != null) {
         for(int i = 0; i < 2; ++i) {
            if (this.widgets[i] != null) {
               int height = this.widgets[i].getPreferredRowHeight(inRow);
               if (height > maxHeight) {
                  maxHeight = height;
               }
            }
         }
      }

      return maxHeight;
   }

   public class ColumnWidget extends AbstractCellEditor implements TableCellRenderer, TableCellEditor {
      int column;
      TableModel tableModel;
      ArrayList<ModelCreationTable.ColumnWidget.Panel> panelArray;

      public ColumnWidget(TableModel inModel, int inColumn) {
         this.column = inColumn;
         this.tableModel = inModel;
         int modelCount = inModel.getRowCount();
         this.panelArray = new ArrayList(modelCount);

         for(int row = 0; row < modelCount; ++row) {
            this.panelArray.add(new ModelCreationTable.ColumnWidget.Panel(row, inColumn, this.tableModel, this.tableModel.getValueAt(row, inColumn)));
         }

      }

      public Component getTableCellRendererComponent(JTable table, Object color, boolean isSelected, boolean hasFocus, int row, int column) {
         return (Component)this.panelArray.get(row);
      }

      public void adjustRowCount(int inRowCount) {
         int row;
         for(row = this.panelArray.size() - 1; row >= inRowCount; --row) {
            this.panelArray.remove(row);
         }

         for(row = this.panelArray.size(); row < inRowCount; ++row) {
            this.panelArray.add(new ModelCreationTable.ColumnWidget.Panel(row, this.column, this.tableModel, this.tableModel.getValueAt(row, this.column)));
         }

      }

      public int getPreferredRowHeight(int inRow) {
         return ((ModelCreationTable.ColumnWidget.Panel)this.panelArray.get(inRow)).getPreferredHeight();
      }

      public Object getCellEditorValue() {
         return new Integer(1);
      }

      public Component getTableCellEditorComponent(JTable table, Object value, boolean isSelected, int row, int column) {
         return (Component)this.panelArray.get(row);
      }

      public class Panel extends JPanel implements FocusListener {
         private JTextField textField;
         private int row;
         private int column;
         private TableModel tableModel;

         public Panel(int inRow, int inColumn, TableModel inTableModel) {
            this.row = inRow;
            this.column = inColumn;
            this.tableModel = inTableModel;
            this.setLayout(new BoxLayout(this, 1));
            this.setOpaque(true);
            this.setBackground(Color.white);
            this.add(this.textField = new JTextField(3));
            this.textField.addFocusListener(this);
         }

         public Panel(int inRow, int inColumn, TableModel inTableModel, Object inValue) {
            this(inRow, inColumn, inTableModel);
            this.textField.setText(inValue.toString());
         }

         public int getPreferredHeight() {
            return this.getPreferredSize().height;
         }

         public void focusGained(FocusEvent e) {
         }

         public void focusLost(FocusEvent e) {
            this.tableModel.setValueAt(this.textField.getText(), this.row, this.column);
         }
      }
   }
}
